src/libstd/net/tcp.rs

```diff
@@ -498,18 +498,47 @@ impl TcpStream {
 
     /// Moves this TCP stream into or out of nonblocking mode.
     ///
-    /// On Unix this corresponds to calling fcntl, and on Windows this
-    /// corresponds to calling ioctlsocket.
+    /// This will result in `read`, `write`, `recv` and `send` operations
+    /// becoming nonblocking, i.e. immediately returning from their calls.
+    /// If the IO operation is successful, `Ok` is returned and no further
+    /// action is required. If the IO operation could not be completed and needs
+    /// to be retried, an error with kind [`io::ErrorKind::WouldBlock`] is
+    /// returned.
+    ///
+    /// On Unix platforms, calling this method corresponds to calling `fcntl`
+    /// `FIONBIO`. On Windows calling this method corresponds to calling
+    /// `ioctlsocket` `FIONBIO`.
     ///
     /// # Examples
     ///
+    /// *Note: the usage of [busy-looping] in the following example is used
```

Comment:
I feel kind of conflicted about these examples. We don't want to get too far down the epoll/IOCP rabbit hole, but it seems pretty bad for the example here to be a thing you should never do. We could maybe have some fake `wait_for_more_data()` function and add something like "typically implemented via platform-specific APIs such as epoll or IOCP" into the docs?

Reply:
<details><summary>is this sorta what you mean?</summary>

```diff
diff --git a/src/libstd/net/tcp.rs b/src/libstd/net/tcp.rs
index 51269d57e6..946584bcf2 100644
--- a/src/libstd/net/tcp.rs
+++ b/src/libstd/net/tcp.rs
@@ -529,7 +529,10 @@ impl TcpStream {
     /// let num_bytes_read = loop {
     ///     match stream.read_to_end(&mut buf) {
     ///         Ok(b) => break b,
-    ///         Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => (),
+    ///         Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+    ///             wait_for_fd(); // typically implemented via platform-specific
+    ///                            // APIs such as epoll or IOCP
+    ///         }
     ///         Err(e) => panic!("encountered IO error: {}", e),
     ///     };
     /// };
@@ -839,13 +842,20 @@ impl TcpListener {
     /// for stream in listener.incoming() {
     ///     let mut stream = match stream {
     ///         Ok(s) => s,
-    ///         Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => continue,
+    ///         Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+    ///             wait_for_more_data(); // typically implemented via platform-specific
+    ///                                   // APIs such as epoll or IOCP
+    ///             continue;
+    ///         }
     ///         Err(e) => panic!("encountered IO error: {}", e),
     ///     };
     ///     loop {
     ///         match stream.write_all(b"foo") {
     ///             Ok(()) => break,
-    ///             Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => (),
+    ///             Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+    ///                 wait_for_fd(); // typically implemented via platform-specific
+    ///                                // APIs such as epoll or IOCP
+    ///             }
     ///             Err(e) => panic!("encountered IO error: {}", e),
     ///         };
     ///     }
```
<details>

Reply:
Yeah, something like that seems better than nothing.